File: 1876-map-of-highest-peak/1876-map-of-highest-peak.js
/**
 * @param {number[][]} isWater
 * @return {number[][]}
 */

 var highestPeak = function (isWater) {
    const m = isWater.length;
    const n = isWater[0].length;

    const height = Array.from({ length: m }, () => Array(n).fill(-1));
    const queue = [];
    
    // Initialize the queue with all water cells
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (isWater[i][j] === 1) {
                height[i][j] = 0;
                queue.push([i, j]); // Push water cells as the starting points
            }
        }
    }
    
    const directions = [
        [0, 1],  // East
        [1, 0],  // South
        [0, -1], // West
        [-1, 0]  // North
    ];
    
    let front = 0; // Efficient deque-like traversal

    // Perform BFS
    while (front < queue.length) {
        const [x, y] = queue[front++]; // Simulate deque using `front` index

        for (const [dx, dy] of directions) {
            const nx = x + dx;
            const ny = y + dy;

            // Check if the neighbor is within bounds and not visited
            if (nx >= 0 && nx < m && ny >= 0 && ny < n && height[nx][ny] === -1) {
                height[nx][ny] = height[x][y] + 1;
                queue.push([nx, ny]); // Add new cell to the queue
            }
        }
    }
    
    return height;
};

var highestPeak1 = function (isWater) {
    const m = isWater.length;
    const n = isWater[0].length;

    const height = Array.from({ length: m }, () => Array(n).fill(-1));
    const queue = [];

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (isWater[i][j] === 1) {
                height[i][j] = 0;
                queue.push([i, j]);
            }
        }
    }

    //console.log("Q: ", queue);

    const directions = [
        [0, 1],
        [1, 0],
        [0, -1],
        [-1, 0]
    ];

    // Perform BFS
    while (queue.length > 0) {
        const [x, y] = queue.shift();
        for (const [dx, dy] of directions) {
            const nx = x + dx;
            const ny = y + dy;

            // Check if the neighbours are within the bounds and not visited
            if (nx >= 0 && nx < m && ny >= 0 && ny < n && height[nx][ny] === -1) {
                //console.log("inside");
                height[nx][ny] = height[x][y] + 1;
                queue.push([nx, ny]);
            }
        }
    }

    //console.log("H:", height);

    return height;
};

File: 1876-map-of-highest-peak/1876-map-of-highest-peak.py
class Solution:
    def highestPeak(self, isWater: List[List[int]]) -> List[List[int]]:
        row_len = len(isWater)
        col_len = len(isWater[0])

        visited = [[0] * col_len for _ in range(row_len)]
        dist = [[0] * col_len for _ in range(row_len)]

        queue = deque()

        for r in range(row_len):
            for c in range(col_len):
                if isWater[r][c] == 1:
                    queue.append((r, c, 0))
                    visited[r][c] = 1
        
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        while queue:
            r, c, steps = queue.popleft()
            dist[r][c] = steps

            for dr, dc in directions:
                nr = r + dr
                nc = c + dc

                if 0 <= nr < row_len and 0 <= nc < col_len and visited[nr][nc] != 1:
                    queue.append((nr, nc, steps + 1))
                    visited[nr][nc] = 1

        return dist